src/core/lib/promise/arena_promise.h

```diff
@@ -149,10 +149,10 @@ class ArenaPromise {
   ArenaPromise() = default;
 
   // Construct an ArenaPromise that will call the given callable when polled.
-  template <typename Callable>
-  ArenaPromise(Arena* arena, Callable&& callable)
+  template <typename Callable, typename Ignored = absl::enable_if_t<!std::is_same<Callable, ArenaPromise>::value>>
```

Comment:
C++ nonsense... by default this template would override the `ArenaPromise&&` constructor and move would break.
This makes it an error to call this particular constructor with a type that is ArenaPromise, which then uses SFINAE to choose the *other* constructor and get move semantics back.

Reply:
Ah I see, thanks.